### ISSUE TITLE
Module Reorganization

### ISSUE THREAD

**Issue by kevinbeaty** [transduce/transduce#18]
Evaluate reorganization of `base` into `core` and `transducers` or anything else that makes sense...

Split from #13

**Comment by kevinbeaty** [comment]
From @tgriesser  on #13 

So I was actually going to ask about the composition of base. I actually don't see too much wrong with having a lot of things on the root namespace a-la underscore, however from a require path standpoint it might make more sense to actually split functionality into a "core" (reduce, transduce, iterator, isReduced, unreduced, completing) and then into a transducers directory, since the individual transducers are for the most part stand alone.

Right now there's a lot of circular dependencies between the different core functionality so there's actually nothing gained by modularizing and IMO it's a bit confusing to have a separate namespace for some of the transformer / iterator pieces that are used frequently.

Keeping things like transduce and reduce in the same file would also allow you to do take some small shortcuts like:

```javascript
function _reduce(xf, init, coll) {
  if(isArray(coll)){
    return arrayReduce(xf, init, coll)
  }
  if(isFunction(coll.reduce)){
    return methodReduce(xf, init, coll)
  }
  return iteratorReduce(xf, init, coll)
}
function reduce(xf, init, coll){
  if(isFunction(xf)){
    xf = completing(xf)
  }
  if (arguments.length === 2) {
    coll = init
    init = xf.init()
  }
  return _reduce(xf, init, coll)
}
function transduce(t, xf, init, coll) {
  if (isFunction(xf)){
    xf = completing(xf)
  }
  xf = t(xf)
  if (arguments.length === 3) {
    coll = init;
    init = xf.init();
  }
  return _reduce(xf, init, coll)
}
```

With an internal shared _reduce to skip the extra isFunction and arguments.length checks and such in the transduce -> reduce step.

**Comment by kevinbeaty** [comment]
Yeah, I was thinking about separating the transducers from the other base/core functionality.  I really do like requiring the functions directly though, especially when using browserify.  I think that is one thing that distinguishes this library from the more popular ones. I don't think there is any harm in including "private" functions like `_reduce` alongside the "public" functions (distinguished by `_` prefix). 

 I tried to avoid circular dependencies, can you point me to a few examples?

I've considered flattening everything into one namespace as well ... but things like `join` can mean different things in different contexts  which is why I didn't do that initially.  (I might want to add a `join` to work with `by` from #11 but haven't thought that through yet).  Still might be a good way to go though.

@RGBBoy Any thoughts on reorganization? I know you would like to maintain the explicit require for builds, and I promised you I wouldn't move things around much :)

```javascript
core = {
reduce: function(xf, init?, coll)
transduce: function(t, xf, init?, coll)
eduction: function(t, coll)
into: function(init, t?, coll?)
sequence: function(t, coll) // this should move now to core now that we are using it in eduction
toArray: function(t?, coll?)
toObject: function(t?, coll?)
toString: function(t?, coll?)
compose: function(/*fns*/)
isReduced: function(value)
reduced: function(value, force?)
unreduced: function(value)
// anything else?
}

transducers = {
map: function(f)
filter: function(predicate)
remove: function(predicate)
take: function(n)
takeWhile: function(predicate)
drop: function(n)
dropWhile: function(predicate)
cat: transducer
mapcat: function(f)
partitionAll: function(n)
partitionBy: function(f)
transformStep: function(f)
}
```

**Comment by RGBboy** [comment]
Looks good to me. I am not yet heavily invested in using this library yet so any changes now will not affect me much. I think it is a good idea to do this refactor. It would be great if once you hit v1.0.0 that you follow semver so that future breaking changes will be communicated through the version number.

**Comment by tgriesser** [comment]
> I tried to avoid circular dependencies, can you point me to a few examples?

Sorry not really circular dependencies, but just like shared dependencies... like if you're requiring `reduce` you're always going to get `iterator`, which means you're always going to get `iterable`, etc... so if the goal is to shave browserify bits you're not gaining anything (and actually probably losing a small bit overall)

Whereas it'd be totally possible to have the inverse where `require('transduce/core/reduce')` was internally just proxied to `module.exports = require('../core').reduce`, and then where you'd really win is requiring the individual transducers by hand, since right now requiring base comes with `drop` or `dropWhile` even if I don't use those.

> I've considered flattening everything into one namespace as well ...

Yeah, I would agree this isn't worth it. Just main pieces that are practical 90% of the time (the core stuff, the any top-level transducing fns, and some of the transformer / iterator functionality) would make sense to keep on the default base namespace. Internally they could still be mapped to individual files as it makes sense for browserify - especially the transducers.

I like the idea of having namespaces for main pieces like `array`, `string`, `push`, `math` - I think some of the `transformer` utility fns is what I was thinking of for core, and maybe `iterable` / `iterator` / `isIterator`.

> I think that is one thing that distinguishes this library from the more popular ones. 

I'd say the distinguishing factor is this library is much further along than the others and more well planned out / maintained. Great work.

**Comment by tgriesser** [comment]
Also, while we're at it... the `toArray`, `toString`, `toObject` have become a bit redundant now with the curried into... thoughts on just dropping them outright?

**Comment by kevinbeaty** [comment]
@RGBboy 
> It would be great if once you hit v1.0.0 that you follow semver so that future breaking changes will be communicated through the version number.
Yup, most definitely.  I actually try to bump the minor version with breaking changes even before v1.0.0, but will use semver once we get there.

@tgriesser 
> Sorry not really circular dependencies, but just like shared dependencies... 
OK, I understand now.  Yeah, that's an interesting tradeoff.  I'll have play around with it and give it some  more thought.

> Also, while we're at it... the toArray, toString, toObject have become a bit redundant now with the curried into... thoughts on just dropping them outright?

I debated with myself on whether to add `toString` and `toObject` or remove `toArray` (which was already there).   I wasn't sure if it would be that obvious what curried `into` really bought you so I added those as convenience.   They are redundant though.  I'll probably remove them and beef up the documentation of `into`.

Thanks for the kind words.  I do appreciate the feedback.

**Comment by kevinbeaty** [comment]
- Split `base` into `core` and `common`.  
- Remove `toArray`, `toObject`, `toString`.  
- Moved `iterator/sequence` into `core`

**Comment by kevinbeaty** [comment]
- Use internal `_reduce` and `_transduce` to skip redundant argument checks internally in `_core` if possible.
- Reorganize tests to match reorganized packages.

**Comment by kevinbeaty** [comment]
I believe I'm done.  I've merged #20 and will close this issue pending further discussion.

**Comment by kevinbeaty** [comment]
Version 0.6
